executor/checksum.go

```diff
@@ -0,0 +1,292 @@
+		chk.AppendUint64(4, t.Response.TotalBytes)
+	}
+	e.done = true
+	return nil
+}
+
+func (e *ChecksumTableExec) buildTasks() ([]*checksumTask, error) {
+	var tasks []*checksumTask
+	for id, t := range e.tables {
+		reqs, err := t.BuildRequests(e.ctx)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		for _, req := range reqs {
+			tasks = append(tasks, &checksumTask{id, req})
+		}
+	}
+	return tasks, nil
+}
+
+func (e *ChecksumTableExec) handleResults(results []*checksumResult) {
+	for _, result := range results {
+		table := e.tables[result.TableID]
+		table.HandleResponse(result.Response)
+	}
+}
+
+func (e *ChecksumTableExec) checksumWorker(taskCh <-chan *checksumTask, resultCh chan<- *checksumResult) {
+	for task := range taskCh {
+		result := &checksumResult{TableID: task.TableID}
+		result.Response, result.Error = e.handleChecksumRequest(task.Request)
+		resultCh <- result
+	}
+}
+
+func (e *ChecksumTableExec) handleChecksumRequest(req *kv.Request) (resp *tipb.ChecksumResponse, err error) {
+	ctx := context.TODO()
+	res, err := distsql.Checksum(ctx, e.ctx.GetClient(), req)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	res.Fetch(ctx)
+	defer func() {
+		if err1 := res.Close(); err1 != nil {
+			err = errors.Trace(err1)
+		}
+	}()
+
+	resp = &tipb.ChecksumResponse{}
+
+	for {
+		data, err := res.NextRaw(ctx)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if data == nil {
+			break
+		}
+		checksum := &tipb.ChecksumResponse{}
+		if err = checksum.Unmarshal(data); err != nil {
+			return nil, errors.Trace(err)
+		}
+		updateChecksumResponse(resp, checksum)
+	}
+
+	return resp, nil
+}
+
+type checksumTask struct {
+	TableID int64
+	Request *kv.Request
+}
+
+type checksumResult struct {
+	Error    error
+	TableID  int64
+	Response *tipb.ChecksumResponse
+}
+
+type checksumContext struct {
+	DBInfo    *model.DBInfo
+	TableInfo *model.TableInfo
+	StartTs   uint64
+	Response  *tipb.ChecksumResponse
+}
+
+func newChecksumContext(db *model.DBInfo, table *model.TableInfo, startTs uint64) *checksumContext {
+	return &checksumContext{
+		DBInfo:    db,
+		TableInfo: table,
+		StartTs:   startTs,
+		Response:  &tipb.ChecksumResponse{},
+	}
+}
+
+func (c *checksumContext) BuildRequests(ctx sessionctx.Context) ([]*kv.Request, error) {
+	reqs, err := c.buildTableRequest(ctx)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	for _, indexInfo := range c.TableInfo.Indices {
+		if indexInfo.State != model.StatePublic {
+			continue
+		}
+		req, err := c.buildIndexRequest(ctx, indexInfo)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		reqs = append(reqs, req)
+	}
+
+	return reqs, nil
+}
+
+func (c *checksumContext) buildTableRequest(ctx sessionctx.Context) ([]*kv.Request, error) {
+	checksum := &tipb.ChecksumRequest{
+		StartTs:   c.StartTs,
+		ScanOn:    tipb.ChecksumScanOn_Table,
+		Algorithm: tipb.ChecksumAlgorithm_Crc64_Xor,
+	}
+
+	var ranges []*ranger.NewRange
+	if c.TableInfo.PKIsHandle {
```

Comment:
If the scan order does not matter for checksum, you can just use `FullIntNewRange(false)`.

Reply:
In my understanding, it still needs to handle the unsigned ranges even if it doesn't care about the order.

Reply:
PTAL @winoros